src/dynamics/circle/rotation_number/basic.lean

```diff
@@ -0,0 +1,654 @@
+  dist (g₁ 0) (g₂ 0) < 2 :=
+dist_map_zero_lt_of_semiconj $ semiconj_by_iff_semiconj.1 h
+
+/-!
+### Estimates on `(f^n) x`
+
+If we know that `f x` is `≤`/`<`/`≥`/`>`/`=` to `x + m`, then we have a similar estimate on
+`f^[n] x` and `x + n * m`.
+
+For `≤`, `≥`, and `=` we formulate both `of` (implication) and `iff` versions because implications
+work for `n = 0`. For `<` and `>` we formulate only `iff` versions.
+-/
+
+lemma iterate_le_of_map_le_add_int {x : ℝ} {m : ℤ} (h : f x ≤ x + m) (n : ℕ) :
+  f^[n] x ≤ x + n * m :=
+by simpa only [nsmul_eq_mul, add_right_iterate]
+using (f.commute_add_int m).iterate_le_of_map_le f.monotone (monotone_id.add_const m) h n
+
+lemma le_iterate_of_add_int_le_map {x : ℝ} {m : ℤ} (h : x + m ≤ f x) (n : ℕ) :
+  x + n * m ≤ (f^[n]) x :=
+by simpa only [nsmul_eq_mul, add_right_iterate]
+using (f.commute_add_int m).symm.iterate_le_of_map_le (monotone_id.add_const m) f.monotone h n
+
+lemma iterate_eq_of_map_eq_add_int {x : ℝ} {m : ℤ} (h : f x = x + m) (n : ℕ) :
+  f^[n] x = x + n * m :=
+by simpa only [nsmul_eq_mul, add_right_iterate]
+using (f.commute_add_int m).iterate_eq_of_map_eq n h
+
+lemma iterate_pos_le_iff {x : ℝ} {m : ℤ} {n : ℕ} (hn : 0 < n) :
+  f^[n] x ≤ x + n * m ↔ f x ≤ x + m :=
+by simpa only [nsmul_eq_mul, add_right_iterate]
+using (f.commute_add_int m).iterate_pos_le_iff_map_le f.monotone (strict_mono_id.add_const m) hn
+
+lemma iterate_pos_lt_iff {x : ℝ} {m : ℤ} {n : ℕ} (hn : 0 < n) :
+  f^[n] x < x + n * m ↔ f x < x + m :=
+by simpa only [nsmul_eq_mul, add_right_iterate]
+using (f.commute_add_int m).iterate_pos_lt_iff_map_lt f.monotone (strict_mono_id.add_const m) hn
+
+lemma iterate_pos_eq_iff {x : ℝ} {m : ℤ} {n : ℕ} (hn : 0 < n) :
+  f^[n] x = x + n * m ↔ f x = x + m :=
+by simpa only [nsmul_eq_mul, add_right_iterate]
+using (f.commute_add_int m).iterate_pos_eq_iff_map_eq f.monotone (strict_mono_id.add_const m) hn
+
+lemma le_iterate_pos_iff {x : ℝ} {m : ℤ} {n : ℕ} (hn : 0 < n) :
+  x + n * m ≤ (f^[n]) x ↔ x + m ≤ f x :=
+by simpa only [not_lt] using not_congr (f.iterate_pos_lt_iff hn)
+
+lemma lt_iterate_pos_iff {x : ℝ} {m : ℤ} {n : ℕ} (hn : 0 < n) :
+  x + n * m < (f^[n]) x ↔ x + m < f x :=
+by simpa only [not_le] using not_congr (f.iterate_pos_le_iff hn)
+
+lemma mul_floor_map_zero_le_floor_iterate_zero (n : ℕ) : ↑n * ⌊f 0⌋ ≤ ⌊(f^[n] 0)⌋ :=
+begin
+  rw [le_floor, int.cast_mul, int.cast_coe_nat, ← zero_add ((n : ℝ) * _)],
+  apply le_iterate_of_add_int_le_map,
+  simp [floor_le]
+end
+
+/-!
+### Definition of translation number
+-/
+noncomputable theory
+
+/-- An auxiliary sequence used to define the translation number. -/
+def transnum_aux_seq (n : ℕ) : ℝ := (f^(2^n)) 0 / 2^n
+
+/-- The translation number of a `circle_deg1_lift`, $τ(f)=\lim_{n→∞}\frac{f^n(x)-x}{n}$. We use
+an auxiliary sequence `\frac{f^{2^n}(0)}{2^n}` to define `τ(f)` because some proofs are simpler
+this way. -/
+def translation_number : ℝ :=
+lim ((at_top : filter ℕ).map f.transnum_aux_seq)
```

Comment:
Wow, this is gross. For liminfs, we have `Liminf` (for a filter) and `liminf` (for a function). Do you think it would be a good idea to do the same for `Lim` and `lim`?

Reply:
I can do it in another PR.